ostd/libs/linux-bzimage/setup/src/x86/amd64_efi/decoder.rs

```diff
@@ -0,0 +1,55 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! This module is used to decompress payload.
+
+extern crate alloc;
+
+pub use alloc::vec::Vec;
+use core::convert::TryFrom;
+
+use core2::io::Read;
+use libflate::{deflate, gzip, zlib};
+
+enum MagicNumber {
+    Elf,
+    Gzip,
+    Zlib,
+    Deflate,
+}
+
+impl TryFrom<&[u8]> for MagicNumber {
+    type Error = ();
+
+    fn try_from(slice: &[u8]) -> Result<Self, Self::Error> {
+        match *slice {
+            [0x7F, 0x45, 0x4C, 0x46] => Ok(Self::Elf),
+            [0x1F, 0x8B] => Ok(Self::Gzip),
+            [0x78, 0x9C] => Ok(Self::Zlib),
+            _ => Ok(Self::Deflate),
+        }
+    }
+}
```

Comment:
I think it doesn't work.

Checkout:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6ab5d6b32261c1a893455714bf1921d3